Java generic bounded types in method signatures

owen-mc-diffblue opened the pull request:
This is based on top of #2231. I will rebase when that PR is merged.

When we find generic bounded types in method signatures we give up and fall back to using the type from the descriptor. This means we get the bound correct.

@smowton and @thk123 please review.

smowton reviewed (approved):
Fine with me if the block-commenting thing is consistent with how their generic tests usually work. Needs a TG bump.

owen-mc-diffblue commented:
TG bump: https://github.com/diffblue/test-gen/pull/1851

owen-mc-diffblue commented:
I've rebased now #2231 has been merged and updated the TG bump

peterschrammel commented on jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -68,6 +68,25 @@ SCENARIO(
+          const typet &param_type = param.type();
+          REQUIRE(param_type.id() == ID_pointer);
+          REQUIRE(param_type.subtype().id() == ID_symbol);
+//          REQUIRE(
```
remove commented code or use `#if 0` AND add a comment why you deactivate the code (several occurrences)

majakusber commented on jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -68,6 +68,28 @@ SCENARIO(
+          // Not currently working as the bound on type variable is ignored
+          REQUIRE(
+            id2string(to_symbol_type(param_type.subtype()).get_identifier()) ==
+            "java::java.lang.Number");
```
Shouldn't this work? By using descriptor rather than signature, the function is non-generic with argument `x` of type `Number` (the `method_name` above suggests so too). Before your changes, `x` was a generic parameter `T` with bound `Object`. In the ideal (future) case, when we do parse bounds, `x` will be generic parameter `T` with bound `Number`. Do I understand this correctly?

This applies to the other tests too.

owen-mc-diffblue replied:
In this case it doesn't work because the generic type and its bound are defined on the class , not on the method, so the changes I've made don't affect it, and it uses the method signature rather than the method descriptor. Apart from that your understanding is correct.

majakusber commented on jbmc/unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -68,6 +68,28 @@ SCENARIO(
+          const typet &param_type = param.type();
+          REQUIRE(param_type.id() == ID_pointer);
+          REQUIRE(param_type.subtype().id() == ID_symbol);
+#if 0
```
Rather than `if 0`, I would just add a comment with a TODO referring to the relevant ticket number that should fix the bounds - TG-1286 perhaps. Similarly in the other tests.

owen-mc-diffblue replied:
Okay. I was trying to save someone time in future working out what tests to write when TG-1286 is fixed.

majakusber reviewed (commented):
This could use a bit more clarification - is it true that before, when seeing bounds, we would keep the function/argument generic but set the bound to `Object` (which is not correct)? And now, by reverting back to descriptor, we make it non-generic and set it's type to the bound (correct but conservative)? This is a correction but not yet a solution to parsing bounds, so I would refer to the relevant ticket in comments.

Commit:
Give up parsing generic method signature with bound

Bounds on generic type variables are not currently parsed. Previously
we were assuming a bound of java.lang.Object. This caused inconsistent
types in goto programs. When we find a bounded type variable in a
function signature we now give up on trying to parse the generic types
and fall back to using the descriptor, which knows about the bounds
and gives us type consistency. The problem still occurs when the bounded
type variable is defined in a class signature.

owen-mc-diffblue commented:
@svorenova, @peterschrammel please re-review.

owen-mc-diffblue commented:
TG bump has passed, so this can be merged when it has enough reviews

peterschrammel reviewed (approved):
Maybe @thk123 has an idea whether there is a 'knownbug' mechanism in catch.

smowton commented:
There is (`[!shouldfail]`), but it applies to whole scenarios / test cases, not individual sections / asserts. That problem has not yet been satisfactorily addressed under Catch AFAIK. Therefore I'll merge this as-is for now, then if anyone thinks of a better answer to knownbug / Catch they should submit a general cleanup PR for it.
